sdk/identity/azure-identity/src/private/token_cache.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Token cache.
+ */
+
+#pragma once
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <functional>
+#include <string>
+
+namespace Azure { namespace Identity { namespace _detail {
+  /**
+   * @brief Implements an access token cache.
+   *
+   */
+  class TokenCache final {
+  private:
+    TokenCache() = delete;
+    ~TokenCache() = delete;
+
+  public:
+    /**
+     * @brief Attempts to get token from cache, and if not found, gets the token using the function
+     * provided, caches it, and returns its value.
+     *
+     * @param tenantId Azure Tenant ID.
+     * @param clientId Azure Client ID.
+     * @param authorityHost Authentication authority URL.
+     * @param scopes Authentication scopes.
+     *
+     * @return Authentication token.
+     *
+     * @throw Azure::Core::Credentials::AuthenticationException Authentication error occurred.
+     */
+    static Core::Credentials::AccessToken GetToken(
+        std::string const& tenantId,
+        std::string const& clientId,
+        std::string const& authorityHost,
+        std::string const& scopes,
+        std::function<Core::Credentials::AccessToken()> const& getNewToken);
+
+    /**
+     * @brief Provides access to internal aspects of the cache as a test hook.
+     *
+     */
+    class Internals;
+
+#if defined(TESTING_BUILD)
+    /**
+     * @brief Clears token cache. Intended to only be used in tests.
+     *
+     */
+    static void Clear();
+#endif
+  };
```

Comment:
I don't think this is a good idea. The code we test would be different from the code we ship.

Reply:
How about move this into `_detail` namespace?

Reply:
This is already in `_detail` namespace, and it is a header under `/src`, so no way for the users to see it.
I agree with the idea as a whole, but realistically, this is the least of all evils.
This is commenting out a static function.
The function is supposed to be used only in tests. It is in a `_detail` class.
But if we comment it out, it is not available to SDK code, only to tests. And should one use it in the SDK code, it will be immediately clear that it should not have been used, and linker error will occur, forcing the SDK programmer to not use it, as intended.

Reply:
ok, it's your call.